### PR TITLE
Systematize ipfs-git parsing a bit

### PR DESCRIPTION
**Pull request by Ericson2314**
Scaled back version of #58.

### CODE REVIEW

File: src/libstore/content-address.hh [hunk 1]
```diff
@@ -158,14 +154,16 @@ struct IPFSInfo {
     PathReferences<IPFSRef> references;
 };
 
+typedef std::variant<
+    TextInfo,
+    FixedOutputInfo,
+    IPFSInfo,
+    IPFSHash
+> ContentAddressWithoutName;
```
Comment (matthewbauer) [inline src/libstore/content-address.hh:162]: we should get a better name for this

File: src/libstore/content-address.cc [hunk 2]
```diff
@@ -224,12 +241,12 @@ void to_json(nlohmann::json& j, const LegacyContentAddress & ca) {
                 { "hash", foh.hash.to_string(Base32, false) },
             };
         },
-        [](IPFSHash foh) {
+        [](IPFSHash ih) {
             return nlohmann::json {
                 { "type", "ipfs" },
-                { "hash", foh.hash.to_string(Base32, false) },
+                { "hash", ih.to_string() },
```
Comment (matthewbauer) [inline src/libstore/content-address.cc:247]: doesn't from_json need to update to reflect this as well?

File: src/libstore/content-address.cc [hunk 3]
```diff
@@ -98,116 +102,129 @@ std::string renderLegacyContentAddress(std::optional<LegacyContentAddress> ca) {
...
+        auto method = parseFileIngestionMethod_(rest);
+        auto hashType = parseHashType_(rest);
+        info = FixedOutputInfo {
+            {
+                .method = method,
+                .hash = Hash::parseNonSRIUnprefixed(rest, std::move(hashType)),
+            },
+            refs,
+        };
+    } else if (tag == "ipfs-git") {
```
Comment (matthewbauer) [inline src/libstore/content-address.cc:216]: no need to restrict ourselves to git here

File: src/libstore/content-address.hh (hunk 1 above)
Comment (matthewbauer) [reply src/libstore/content-address.hh:162]: Naming for these could be:

- ContentAddress becomes StorePathDescriptor
- ContentAddressWithoutName becomes ContentAddressWithReferences
- LegacyContentAddress becomes ContentAddress

Comment (Ericson2314) [reply src/libstore/content-address.hh:162]: I like `StorePathDescriptor`

File: src/libstore/content-address.cc [hunk 4]
```diff
@@ -98,112 +102,129 @@ std::string renderLegacyContentAddress(std::optional<LegacyContentAddress> ca) {
...
-                .name = name,
-                .info = hash
-            };
-        } else throw Error("unknown content address type for '%s'", rawCa);
+
+    std::string_view name;
+    std::string_view tag;
+    {
+        auto optName = splitPrefixTo(rest, ':');
+        auto optTag = splitPrefixTo(rest, ':');
```
Comment (matthewbauer) [inline src/libstore/content-address.cc:158]: I would expect tag to before anything else

Comment (Ericson2314) [reply src/libstore/content-address.cc:158]: Well I wanted to keep the name and ca-with-refs separate. I like your idea of making it separate args, eventually.

**Pull request merged**